pinot-integration-tests/src/test/java/com/linkedin/pinot/integration/tests/BaseClusterIntegrationTest.java

```diff
@@ -78,303 +99,478 @@
+        }
         JSONObject firstAggregationResult = aggregationResultsArray.getJSONObject(0);
+
         if (firstAggregationResult.has("value")) {
-          LOGGER.debug("Trying to execute sql query: " + sqlQueries.get(0));
-          statement.execute(sqlQueries.get(0));
-          ResultSet rs = statement.getResultSet();
-          LOGGER.debug("Trying to get result from sql: " + rs);
-          // Single value result for the aggregation, compare with the actual value
-          final String bqlValue = firstAggregationResult.getString("value");
-
-          rs.first();
-          final String sqlValue = rs.getString(1);
-
-          LOGGER.debug("bql value: " + bqlValue);
-          LOGGER.debug("sql value: " + sqlValue);
-          long compareSqlValue = -1;
-          long compareBqlValue = -1;
-
-          if (bqlValue != null && sqlValue != null) {
-            // H2 returns float and double values in scientific notation. Convert them to plain notation first..
-            try {
-              compareSqlValue = new BigDecimal(sqlValue).longValue();
-              compareBqlValue = new BigDecimal(bqlValue).longValue();
-            } catch (NumberFormatException e) {
-              LOGGER.warn("Ignoring number format excection in " + sqlValue);
-              compareBqlValue=-2; // So comparison will fail below
+          // Aggregation results.
+
+          // Check over all aggregation functions.
+          for (int i = 0; i < numAggregationResults; i++) {
+            // Get expected value for the aggregation.
+            h2statement.execute(sqlQueries.get(i));
+            ResultSet sqlResultSet = h2statement.getResultSet();
+            sqlResultSet.first();
+            String sqlValue = sqlResultSet.getString(1);
+
+            // If SQL value is null, it means no record selected in H2.
+            if (sqlValue == null) {
+              // Check number of documents scanned is 0.
+              if (numDocsScanned != 0) {
+                String failureMessage =
+                    "No record selected in H2 but number of records selected in Pinot: " + numDocsScanned;
+                failure(pqlQuery, sqlQueries, failureMessage);
+              }
+              // Skip further comparison.
+              return;
             }
-          }
 
-          if (GATHER_FAILED_QUERIES) {
-            if (!EqualityUtils.isEqual(compareBqlValue, compareSqlValue)) {
-              saveFailedQuery(pqlQuery, sqlQueries, "Values did not match for query " + pqlQuery + ", expected "
-                  + sqlValue + ", got " + bqlValue);
+            // Get actual value for the aggregation.
+            String pqlValue = aggregationResultsArray.getJSONObject(i).getString("value");
+
+            // Fuzzy compare expected value and actual value.
+            double expectedValue = Double.parseDouble(sqlValue);
+            double actualValue = Double.parseDouble(pqlValue);
+            if (Math.abs(actualValue - expectedValue) >= 1.0) {
```

Comment:
We have an option to log all failed queries, so in that case we don't want early termination.